### PR TITLE
[eslint] propagate config Rules to `overrides` property

### PR DESCRIPTION
**Pull request by kazarmy**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://eslint.org/docs/latest/user-guide/configuring/configuration-files#how-do-overrides-work "Override blocks can contain any configuration options that are valid in a regular config, ..." 
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

This change enables ESLint rules documentation on mouse hover for the rules in the `overrides` property.

### CODE REVIEW

**Comment by JounQin** [comment]
The overrides rules can be different from the base, right?

**Comment by kazarmy** [comment]
Yes, different rule sets can be enabled in an overrides entry -- they can even have different parsers. For example, for `*.ts` files, it would be `@typescript-eslint` rules with the `@typescript-eslint/parser` parser. However, I think it's better to have some typings available on the `overrides` property rather than none at all. At least mistakes with ESLint rule parameter names can be detected early.

**Comment by kazarmy** [comment]
And from my testing, it does seem viable to have something like:

```ts
const config: Linter.Config<ESLintRules & MyTypeScriptESLintRules> = {
  // ...
};
```
if the rule typings need to be extended further.

**Comment by kazarmy** [comment]
Ping @pmdartus @j-f1 @saadq @JasonHK @bradzacher @JounQin ... apparently @typescript-bot doesn't do reminders for this particular situation

**Comment by JounQin** [comment]
> And from my testing, it does seem viable to have something like:
> 
> ```ts
> const config: Linter.Config<ESLintRules, MyTypeScriptESLintRules> = {
>   // ...
> };
> ```
> 
> if the rule typings need to be extended further.

How about add another generic parameter?

**Comment by kazarmy** [comment]
Well there can be more than two sets of rule typings (e.g. for `eslint`, `@typescript-eslint` and `jsonc`) that need to be concatenated together so I think using `&` as seen [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62356#issuecomment-1255152821) instead of adding generic parameters would be better.

**Comment by kazarmy** [comment]
Anyway whether or not `&` or `,` is better, the config `Rules` typings should still be propagated to the `overrides` property so that some rules typings support is available even if not all.

**Comment by kazarmy** [comment]
I would be really surprised if this change causes massive ecosystem effects.

**Comment by andrewbranch** [comment]
@JounQin Any other thoughts? This seems like an improvement to me.

File: types/eslint/index.d.ts [hunk 1]
```diff
@@ -725,7 +725,7 @@ export namespace Linter {
         extends?: string | string[] | undefined;
         globals?: { [name: string]: boolean | "readonly" | "readable" | "writable" | "writeable" } | undefined;
         noInlineConfig?: boolean | undefined;
-        overrides?: ConfigOverride[] | undefined;
+        overrides?: Array<ConfigOverride<Rules>> | undefined;
```
Comment (JounQin) [inline types/eslint/index.d.ts:728]: I still think something like the following would make more sense?

```ts
interface BaseConfig<Rules extends RulesRecord = RulesRecord, OverrideRules extends RulesRecord = Rules> extends HasRules<Rules> {
  overrides?: Array<ConfigOverride<OverrideRules>> | undefined
}
```

Comment (kazarmy) [reply types/eslint/index.d.ts]: Ok done 614bb973554875f10ddf1cbcaadae2d283f54c37

**Review by JounQin — APPROVED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by kazarmy** [comment]
Ready to merge

**Pull request merged**